### ISSUE TITLE
Support for Cloudflare Workers

### ISSUE THREAD

**Issue by davidmytton** [arcjet/arcjet-js#759]
Allow us to support applications built on Cloudflare workers. There's a lot of useful data in the request context [`cf` property](https://developers.cloudflare.com/workers/runtime-apis/request/#the-cf-property-requestinitcfproperties) we could make use of.

**Comment by blaine-arcjet** [comment]
This is blocked by Connect RPC's requirement to construct a client with a baseUrl. We want to avoid reconstructing a client on each request to the worker.

**Comment by blaine-arcjet** [comment]
Related: Connect doesn't work well on Cloudflare, see https://github.com/connectrpc/connect-es/issues/577

**Comment by davidmytton** [comment]
I spent a bit of time researching where we're at with support for Cloudflare workers. 

> This is blocked by Connect RPC's requirement to construct a client with a baseUrl. We want to avoid reconstructing a client on each request to the worker.

Seems there's a few reasons for this:

1. We want to load our WASM module into the worker outside of the request context so it can be instantiated once rather than on every request. This is why we recommend creating the Arcjet client outside of the request handler for our other adapters.
2. With https://github.com/arcjet/arcjet-js/pull/1201 we'd like to have a persistent connection to improve latency across requests.
3. The DX is better when you can define base rules in a client and then `withRule` any additional rules you might want to combine.

With Cloudflare's [move from Service Workers to ES Modules](https://developers.cloudflare.com/workers/reference/migrate-to-module-workers/), environment variables are no longer provided in the global scope. Instead, they're [passed as a binding to the handler](https://developers.cloudflare.com/workers/reference/migrate-to-module-workers/#environment-variables-in-es-modules-format). Trying to access them as a global will error.

Also, creating the connect client requires a base URL, which we configure dynamically based on the environment variables.

**Loading WASM**

Cloudflare officially supports WASM in workers. They have [some history with slow worker startup](https://community.cloudflare.com/t/fixed-cloudflare-workers-slow-with-moderate-sized-webassembly-bindings/184668), but it seems they've implemented a fix for this. [From 2020](https://community.cloudflare.com/t/fixed-cloudflare-workers-slow-with-moderate-sized-webassembly-bindings/184668/16):

> We found a way to enable V8’s Liftoff+Tier-up that we like. This should go out to production next week. I think it should cut Wasm startup time by a factor of 5-10. (This still doesn’t use code cache, which is a much bigger project that should cut Wasm startup time to almost nothing. No timeline on that yet, but let’s see how much Liftoff helps.)

It would be interesting to test the startup time of our WASM Module. They [must be instantiated in the global scope](https://developers.cloudflare.com/workers/runtime-apis/webassembly/javascript/#use-from-javascript) outside of the handler, which implies some optimization may be going on.

There is also a [worker size limit of 1MB](https://developers.cloudflare.com/workers/platform/limits/#account-plan-limits) for free plans, 10MB for paid plans.

**Persistent connections**

Cloudflare doesn't support persistent outbound connections so we don't get any benefit here.

You can see in [the examples for databases](https://neon.tech/docs/guides/cloudflare-workers#implement-the-worker-script), they create a new connection within the workers handler, then it closes after the request ends.

Given this, we could avoid our usage of the dynamic `baseUrl` and use a static URL when in the workers environment.

I assume we'd use the fetch client from Connect: https://github.com/connectrpc/connect-es/issues/577#issuecomment-1503185372

Did I miss anything or get anything wrong?

**Comment by prometixX** [comment]
is there any update for the cloudflare worker support?

**Comment by davidmytton** [comment]
Hi @prometixX - no updates I'm afraid. We've prioritized other product areas so don't have an ETA.

**Comment by wooorm-arcjet** [comment]
Did some digging into this, and at least got it to work.

* do step 1 through 3 from https://developers.cloudflare.com/workers/get-started/guide/
* `"type": "module",` in `package.json` for ESM
* run `npx wrangler types` for types
* add some tsconfig to check things (example at the bottom)
* add a `.env` with an `ARCJET_KEY=XXX`
* install `@arcjet/headers`, `@arcjet/ip`, `@arcjet/protocol`, `@arcjet/transport`, `arcjet`
* ~apply the patches from GH-5077 locally in `node_modules` (just look for the `package.json` changes, and point them to the `edge-runtime` files)~
* then, in `src/index.js`:

```js
import {env} from "cloudflare:workers";
import {ArcjetHeaders} from '@arcjet/headers'
import {findIp} from '@arcjet/ip'
import {createClient} from '@arcjet/protocol/client.js'
import {createTransport} from '@arcjet/transport'
import arcjetCore, {detectBot} from 'arcjet'

const baseUrl = 'https://decide.arcjet.com'

const client = createClient({
  baseUrl,
  sdkStack: 'FASTIFY', // A small lie.
  sdkVersion: '1.0.0',
  timeout: 500,
  transport: createTransport(baseUrl),
})

const arcjet = arcjetCore({
  characteristics: ['http.request.headers["user-agent"]', "ip.src"],
  client,
  key: env.ARCJET_KEY,
  log: console,
  rules: [detectBot({deny: ['CURL'], mode: 'LIVE'})]
})

export default {
  /**
   * @param {Request} request
   * @param {Record<string, string>} _env
   * @param {ExecutionContext} ctx
   * @returns {Promise<Response>}
   */
  async fetch(request, _env, ctx) {
    const now = Date.now()

    const url = new URL(request.url)

    const decision = await arcjet.protect(
      {
        async getBody() { throw new Error("Not implemented") },
        waitUntil: ctx.waitUntil.bind(ctx)
      },
      {
        cookies: request.headers.get('cookie') || '',
        host: url.host,
        headers: new ArcjetHeaders(request.headers),
        ip: findIp(request, {platform: 'cloudflare'}) || '127.0.0.1',
        method: request.method,
        path: url.pathname,
        protocol: url.protocol,
        query: url.search,
      }
    )

    // console.log(request)
    console.log('Request finished with ' + decision.conclusion + ' in ' + (Date.now() - now) + 'ms')

    // `curl --header "User-Agent: curl/7.79.1" --verbose http://localhost:8787` results in Access denied.
    // (my curl is configured locally to use a different header, hence explicit)
    // Visiting in a browser results in a successful response.
    return decision.conclusion === 'DENY'
      ? new Response('Access denied', {status: 403})
      : new Response('Hello World!')
  },
};
```

Running all that with `npx wrangler dev`, gives denies and allows.
Response times are about 3ms for a deny (as that happens locally) and 50ms for an allow (as that contacts the decide api).

<details><summary>example <code>tsconfig.json</code></summary>

```json
{
  "compilerOptions": {
    "checkJs": true,
    "customConditions": ["development"],
    "declarationMap": true,
    "declaration": true,
    "emitDeclarationOnly": true,
    "exactOptionalPropertyTypes": true,
    "lib": ["es2022"],
    "module": "node16",
    "strict": true,
    "target": "es2022"
  },
  "exclude": ["node_modules/"],
  "include": ["**/*.js", "worker-configuration.d.ts"]
}
```

</details>

**Comment by davidmytton** [comment]
With `aj` created inside the `fetch` handler, doesn't this mean the SDK is instantiated for every request? So we lose the benefits of pre-loaded WASM and the persistent connection to our API.

**Comment by wooorm-arcjet** [comment]
Right. That’s why I wanted to measure the time! 5ms for local deny (which uses WebAssembly), 50ms remote allow! On my computer.

(Though, how long lived are these edge workers? There will be more startups regardless of what we come up with. So we’d want to make that as fast as possible. Right?)

Importantly in this example, compared to your last comment, the `client` is *outside*: the `key` variable comes from the `.env`, which is given per request. That’s why it needs to be re-initialized.

**Comment by wooorm-arcjet** [comment]
Moving the construction of `aj` outside, with a hardcoded key, changes the times to 3ms for a local deny, 45ms for a remote allow

**Comment by The-LukeZ** [comment]
May I come in here and bring this to attention again?

**Comment by wooorm-arcjet** [comment]
It’s now possible to `import {env} from "cloudflare:workers"`

- https://developers.cloudflare.com/workers/runtime-apis/bindings/#importing-env-as-a-global
- https://developers.cloudflare.com/workers/configuration/environment-variables/#import-env-for-global-access

I’ve amended my above example

**Comment by davidmytton** [comment]
Nice! I also had a report from a customer that enabling [node compatibility mode in Workers](https://developers.cloudflare.com/workers/runtime-apis/nodejs/) allows them to use the Arcjet Node SDK with no modifications, although I've not validated that myself yet.

**Comment by davidmytton** [comment]
Arcjet supports Cloudflare Workers for both `protect()` and `guard()`

**What's in place:**

- `@arcjet/transport`, `@arcjet/analyze-wasm`, and `@arcjet/runtime` all ship dedicated `workerd` entry points, so the SDK resolves the right code path when bundled for Workers.
- `@arcjet/ip` exports a [`cloudflare()`](https://github.com/arcjet/arcjet-js/blob/main/ip/cloudflare.ts) helper that reads `CF-Connecting-IP` / `CF-Connecting-IPv6` for client-IP detection — directly addressing the `cf` request data mentioned in this issue.

**Verified end-to-end on Workers (via miniflare):**

- `guard()` — covered by [`arcjet-guard/test/runtime/cloudflare`](https://github.com/arcjet/arcjet-js/tree/main/arcjet-guard/test/runtime/cloudflare).
- `protect()` (core SDK) — covered by a new runtime test added in #6110. It bundles a Worker through the `workerd` transport + the `analyze-wasm` `workerd` WASM, runs `protect()` against a real mock DecideService, and confirms the request leaves the Worker over the network and returns a decision.

Closing as resolved once #6110 lands. If you hit any issues running on Workers, please open a fresh issue.
